### ISSUE TITLE
deleteById() / deleteAll() not working with indexed properties

### ISSUE THREAD

**Issue by markbigler** [bwaldvogel/mongo-java-server#39]
I have found another bug which prevents me from switching from `fongo` to `mongo-java-server`.

If you modify an indexed property of a document and then delete the document, no new document can be inserted with the value of the originally indexed property.

Here is a test case to reproduce the problem:

``` java
@RunWith(SpringRunner.class)
@SpringBootTest(classes={Issue.TestConfiguration.class})
public class Issue {

	@Autowired private TestRepository repository;

	@Test
	public void testDeleteWithUniqueIndexes() throws Exception {
		TestEntity document = repository.save(new TestEntity("DOC_1", "Text1"));

		// update value of indexed property
		document.setText("Text1 (updated)");
		repository.save(document);

		// delete document (deleteAll() does not work either)
		repository.deleteById("DOC_1");

		// duplicate key error
		repository.save(new TestEntity("DOC_1", "Text1"));
	}


	@Configuration
	@EnableMongoRepositories(basePackageClasses={TestRepository.class})
	protected static class TestConfiguration {
		@Bean
		public MongoTemplate mongoTemplate(MongoClient mongoClient) {
			return new MongoTemplate(mongoDbFactory(mongoClient));
		}

		@Bean
		public MongoDbFactory mongoDbFactory(MongoClient mongoClient) {
			return new SimpleMongoDbFactory(mongoClient, "test");
		}

		@Bean(destroyMethod="shutdown")
		public MongoServer mongoServer() {
			MongoServer mongoServer = new MongoServer(new MemoryBackend());
			mongoServer.bind();
			return mongoServer;
		}

		@Bean(destroyMethod="close")
		public MongoClient mongoClient(MongoServer mongoServer) {
			return new MongoClient(new ServerAddress(mongoServer.getLocalAddress()));
		}
	}
}
```

``` java
@Document(collection="test")
public class TestEntity {
	@Id private String id;
	@Indexed(unique=true) private String text;
	public TestEntity() {
	}
	public TestEntity(String id, String text) {
		this.id = id;
		this.text = text;
	}
	public String getId() {
		return id;
	}
	public String getText() {
		return text;
	}
	public void setText(String text) {
		this.text = text;
	}
}
```

``` java
public interface TestRepository extends MongoRepository<TestEntity, String> {
}
```

**Comment by markbigler** [comment]
Some more assertions / debug output:

``` java
@Test
public void testDeleteWithUniqueIndexes() throws Exception {
	...
	repository.deleteById("DOC_1");

	// somehow the document is still returned in findAll() ...
	repository.findAll().forEach(e -> {
		System.out.println("Still existing: " + e.getId());
	});
		
	// ... but not with findById()
	assertFalse(repository.findById("DOC_1").isPresent());

	// fails: expected <0> but was <1>
	assertEquals(0, repository.count());
	...
}
```

**Comment by bwaldvogel** [comment]
Thanks, I can confirm the bug. Root cause is a missing update of the secondary index.

**Comment by markbigler** [comment]
Thanks for the fix. The problem with the missing update seems to be solved, but other test cases that were successful before do not work anymore.

I tried to describe the problem with a minimal example and therefore omitted `sparse=true`.

``` java
public class TestEntity {
	@Id private String id;
	@Indexed(unique=true, sparse=true) private String text;

	...
}
```


``` java
@Test
public void testWithUniqueSparseIndex() throws Exception {
	repository.save(new TestEntity("DOC_2", null));
	repository.save(new TestEntity("DOC_3", null)); // E11000 duplicate key error collection: test.test index: text_1 dup key: { : null };
}
```

**Comment by bwaldvogel** [comment]
It’s caused by missing support for sparse indices. I’m currently working on it and hope to get it pushed soon.

**Comment by markbigler** [comment]
> Can you retest on [d94e915](https://github.com/bwaldvogel/mongo-java-server/commit/d94e915f0223d3757cd8dfb518aaa40393b24431)?

All the ~350 tests using MongoDB are working now! 💚

**Comment by bwaldvogel** [comment]
That’s great to hear :grinning:

I’ve just released version `1.11.0`
